framework/src/modules/chain/bft/bft.js

```diff
@@ -0,0 +1,283 @@
+		this.updatePreVotedAndFinalizedHeight();
+
+		return this;
+	}
+
+	/**
+	 * Update pre-votes and pre-commits in reference to particular block header
+	 *
+	 * @param {BlockHeader} lastBlockHeader
+	 * @return {undefined}
+	 */
+	updatePreVotesPreCommits(lastBlockHeader) {
+		// Update applies particularly in reference to last block header in the list
+		const header = lastBlockHeader || this.headers.last;
+
+		// If delegate forged a block with higher or same height previously
+		// that means he is forging on other chain and we don't count any
+		// pre-votes and pre-commits from him
+		if (header.maxHeightPreviouslyForged >= header.height) {
+			return;
+		}
+
+		// Get delegate public key
+		const delegatePublicKey = header.delegatePublicKey;
+
+		// Load or initialize delegate state in reference to current BFT block headers
+		const delegateState = this.state[delegatePublicKey] || {
+			maxPreVoteHeight: 0,
+			maxPreCommitHeight: 0,
+		};
+
+		// Get first block of the round when delegate was active
+		const delegateMinHeightActive =
+			(header.activeSinceRound - 1) * this.ACTIVE_DELEGATES + 1;
+
+		// If delegate is new then first block of the round will be considered
+		// if it forged before then we probably have the last commit height
+		// delegate can't pre-commit and block before above mentioned conditions
+		const minPreCommit = Math.max(
+			delegateMinHeightActive,
+			delegateState.maxPreCommitHeight + 1
+		);
+
+		// Delegate can't pre-commit the blocks on tip of the chain
+		const maxPreCommitHeight = header.height - 1;
+
+		for (let j = minPreCommit; j <= maxPreCommitHeight; j++) {
+			// Add pre-commit if threshold is reached
+			if (this.preVotes[j] >= this.PRE_VOTE_THRESHOLD) {
+				// Increase the pre-commit for particular height
+				this.preCommits[j] = (this.preCommits[j] || 0) + 1;
+
+				// Keep track of the last pre-commit point
+				delegateState.maxPreCommitHeight = j;
+			}
+		}
+
+		// Check between height of first block of the round when delegate was active
+		// Or one step ahead where it forged the last block
+		// Or one step ahead where it left the last pre-vote
+		// Or maximum 3 rounds backward
+		const minPreVoteHeight = Math.max(
+			delegateMinHeightActive,
+			header.maxHeightPreviouslyForged + 1,
+			delegateState.maxPreVoteHeight + 1,
+			header.height - this.PROCESSING_THRESHOLD
+		);
+		// Pre-vote upto current block height
+		const maxPreVoteHeight = header.height;
+		for (let j = minPreVoteHeight; j <= maxPreVoteHeight; j++) {
+			this.preVotes[j] = (this.preVotes[j] || 0) + 1;
+		}
+		// Update delegate state
+		delegateState.maxPreVoteHeight = maxPreVoteHeight;
+
+		// Set the delegate state
+		this.state[delegatePublicKey] = delegateState;
+	}
+
+	/**
+	 * Update the pre-voted confirmed and finalized height
+	 */
+	updatePreVotedAndFinalizedHeight() {
+		if (this.headers.length === 0) {
+			return;
+		}
+
+		const higherPairVoted = Object.entries(this.preVotes)
+			.reverse()
```

Comment:
why reversing here?

Reply:
Because the `entries` returned in format `[[1, 4], [2, 3], [3, 2].....]` while height as the first element of the pair. We need the highest height pair first, since we don't have any `findLast` method on array, so we first reverse the array and then use `find`